design/external-authorization-design.md

```diff
@@ -0,0 +1,715 @@
+  These specific authorization protocols can be implemented in external authorization servers (see discussion below).
+
+## High-Level Design
+A new `ExtensionService` CRD adds a way to represent and track an authorization service.
+This CRD is relatively generic, so that it can be re-used for Envoy rate limiting and logging services.
+The core of the `ExtensionService` CRD is subset of the `projectcontour.v1.HTTPProxy` `Service` specification.
+Re-using the `Service` type allows the operator to specify configuration in familiar and consistent terms, especially TLS configuration.
+
+Note that only the Envoy [GRPC authorization protocol][2] will be supported.
+The GRPC protocol is a superset of the HTTP protocol and requires less configuration.
+The drawback of only supporting the GRPC protocol is that many existing Envoy authorization servers only support the HTTP protocol.
+Until Contour adds support for the v3 Envoy API, only the version 2 of the GRPC authorization protocol can be configured.
+
+Operators can bind an authorization service to a `HTTPProxy` using a new field in the `VirtualHost` struct.
+This field configures which `ExtensionService` resource to use, and allows the operator to set initial authorization policy.
+Authorization policy can also be set on a `Route`, so that application owners can pass metadata and disable authorization on specific routes.
+
+## Detailed Design
+
+### ExtensionService Changes
+
+The `ExtensionService` CRD has API version `projectcontour.io/v1alpha1`.
+
+There are a number of benefits to creating a CRD to represent a supporting service:
+
+- The CRD directly generates an Envoy [Cluster][4] that Contour can use for any purpose.
+  This means that with a single new API, Contour can add support for authorization, rate limiting and logging support services.
+- A CRD gives Contour a way to communicate the operational status of the support service, which was a [desired][9] [goal][7].
+- A CRD allows the team operating the support service and the team operating Contour to collaborate more loosely.
+
+```Go
+// SupportProtocolVersion is the version of the GRPC protocol used
+// to request support services.
+type SupportProtocolVersion string
+
+// SupportProtocolVersion2 requests the "v2" support protocol version.
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type ExtensionService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ExtensionServiceSpec   `json:"spec,omitempty"`
+	Status ExtensionServiceStatus `json:"status,omitempty"`
+}
+
+// ExtensionServiceStatus should follow the pattern being established in [#2642][10].
+// This field will updated by Contour.
+type ExtensionServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
+}
+
+type ExtensionServiceSpec struct {
+	// +required
+	Service contourv1.Service `json:"service"`
```

Comment:
Just thinking out loud, should `Service` be an array? If you did that like HTTPProxy does, you would be able to shift traffic to new instances of an auth backend via weights.

Reply:
Yeh, I did consider that, but you can already handle this use case (to some extent) with services and labels. If this was a slice, then weights and mirroring could be implemented, which would make it the same as `HTTPProxy`.

Reply:
> already handle this use case (to some extent) with services and labels

This was the reason HTTPProxy has it as a slice since labels & selectors aren't a great way to effectively implement a rollout.